### PR TITLE
chore(PlayerScript/Hook) : Add CanSetTradeItem Hook in PlayerScript

### PR DESCRIPTION
**Pull request by Elmsroth**
## Changes Proposed:

This PR proposes changes to:
-  [x] Scripts (bosses, spell scripts, creature scripts).

## Issues Addressed:

- Create a new hook in PlayerScript : CanSetTradeItem
It allows you to filter the item you want to be allowed to be traded.
It goes further than the CanInitTrade existing hook.

## SOURCE:

The changes have been validated through:
_N/A_

## Tests Performed:

This PR has been:
- [x] Tested in-game by the author.
You can use this module to test the hook : https://github.com/Elmsroth/mod-trade-items-filter

## How to Test the Changes:

1. You can use this PlayerScript to check the hook behaviour : https://gist.github.com/Elmsroth/d8bc358791ab6cf7db501b5cad9126ad  (test script not added to PR since it is not required, it would pollute the repo).

## Known Issues and TODO List:

_N/A_

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Scripting/ScriptDefines/PlayerScript.cpp [hunk 1]
```diff
@@ -1508,6 +1508,21 @@ bool ScriptMgr::CanInitTrade(Player* player, Player* target)
...
 
+bool ScriptMgr::CanSetTradeItem(Player* player, Item* tradedItem)
+{
+    auto ret = IsValidBoolScript<PlayerScript>([&](PlayerScript* script)
+        {
+            return !script->CanSetTradeItem(player, tradedItem);
+        });
+
+    if (ret && *ret)
+    {
```
Comment (sudlud) [inline src/server/game/Scripting/ScriptDefines/PlayerScript.cpp:1519]: You can remove the braces here

File: src/server/game/Scripting/ScriptDefines/PlayerScript.h [hunk 2]
```diff
@@ -421,6 +421,16 @@ class PlayerScript : public ScriptObject
 
     [[nodiscard]] virtual bool CanInitTrade(Player* /*player*/, Player* /*target*/) { return true; }
 
+    /**
+     * @brief This hook called just before finishing the handling of the action of a player seting an item in a trade slot
```
Comment (sudlud) [inline src/server/game/Scripting/ScriptDefines/PlayerScript.h:425]: setting

**Comment by Elmsroth** [comment]
@sudlud : Custom module is available for testing the hook : https://github.com/Elmsroth/mod-trade-items-filter

**Comment by Elmsroth** [comment]
@AzerothCoreDev Can anyone merge please ?

**Comment by sudlud** [comment]
Just tested this together with your module.

mod-trade-items-filter.conf.dist
```
ModTradeItemsFilter.Common.ModuleEnabled = 1
ModTradeItemsFilter.Common.ForbiddenItemsIDs = "42"
ModTradeItemsFilter.GameMasters.MinAccountSecurityLevelToByPassForbiddenItems = 4
```

`.additem 42`

with GM account (level 3) trading item with another player
adding the item to trade window immediately cancels the trade and closes the trade window.

So the PR works as intended.

**Review by sudlud — APPROVED** [review]

**Review by pangolp — APPROVED** [review]
Thanks.

**Comment by Elmsroth** [comment]
Thanks @sudlud and @pangolp

**Pull request merged**